### PR TITLE
Master brace expansion

### PR DESCRIPTION
**Pull request by deniszh**
Port of https://github.com/graphite-project/graphite-web/pull/1531 to master branch.

### CODE REVIEW

**Comment by deniszh** [comment]
Maybe can be merged with https://github.com/graphite-project/graphite-web/pull/1290
/cc @cbowman0

**Comment by obfuscurity** [comment]
Can we also include some new tests?

**Comment by deniszh** [comment]
Will do.

**Comment by deniszh** [comment]
Tests are added, looks better now.

**Comment by obfuscurity** [comment]
Thanks for adding the tests. @deniszh if you feel confident about this and #1531, feel free to merge both.

**Comment by deniszh** [comment]
Thanks, merging both.

File: webapp/graphite/finders/__init__.py [hunk 1]
```diff
@@ -41,21 +42,37 @@ def extract_variants(pattern):
...
+    return list(_deduplicate(matching))
+
+
+"""
+    Brace expanding patch for python3 borrowed from:
+    https://bugs.python.org/issue9584
+"""
+def expand_braces(orig):
+    r = r'.*(\{.+?[^\\]\})'
+    p = re.compile(r)
```
Comment (iain-buclaw-sociomantic) [inline webapp/graphite/finders/__init__.py]: Shouldn't this `re.compile` be a global? I would have thought it expensive to run it every time `expand_braces` is called.

File: webapp/graphite/finders/__init__.py [hunk 2]
```diff
@@ -41,21 +42,37 @@ def extract_variants(pattern):
...
+
+    s = orig[:]
```
Comment (iain-buclaw-sociomantic) [inline webapp/graphite/finders/__init__.py]: I don't think this copy is needed, `.replace` doesn't change the contents inplace afaics. (Just a couple of notes for optimization :-)

**Pull request merged**